File: feature/forecast/src/iosMain/swift/presentation/weeklyforecast/TemperatureRangeBar.swift
import SwiftUI

/// Cold-to-warm gradient used by the daily temperature range bars.
private let temperatureGradient = LinearGradient(
    colors: [
        Color(red: 0x4F / 255, green: 0xC3 / 255, blue: 0xF7 / 255),
        Color(red: 0xFF / 255, green: 0xB7 / 255, blue: 0x4D / 255),
    ],
    startPoint: .leading,
    endPoint: .trailing
)

/// Horizontal bar showing where `low`...`high` sits inside the `scaleMin`...`scaleMax` range
/// shared by every day of the forecast.
struct TemperatureRangeBar: View {
    let low: Double
    let high: Double
    let scaleMin: Double
    let scaleMax: Double

    var body: some View {
        let scaleSpan = scaleMax - scaleMin > 0 ? scaleMax - scaleMin : 1
        let startFraction = clamp((low - scaleMin) / scaleSpan, 0, 1)
        let endFraction = clamp((high - scaleMin) / scaleSpan, startFraction, 1)
        // A single-degree day would otherwise collapse to an invisible bar.
        let fillFraction = max(endFraction - startFraction, 0.05)
        let leadingFraction = min(startFraction, 1 - fillFraction)

        GeometryReader { geometry in
            Capsule()
                .fill(temperatureGradient)
                .frame(width: geometry.size.width * fillFraction)
                .offset(x: geometry.size.width * leadingFraction)
        }
        .frame(height: 8)
        .background(Color(.systemFill), in: Capsule())
    }

    private func clamp(_ value: Double, _ lower: Double, _ upper: Double) -> Double {
        min(max(value, lower), upper)
    }
}

File: feature/forecast/src/androidMain/kotlin/com/igorwojda/showcase/feature/forecast/presentation/weeklyforecast/TemperatureRangeBar.kt
package com.igorwojda.showcase.feature.forecast.presentation.weeklyforecast

import androidx.compose.foundation.background
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.Spacer
import androidx.compose.foundation.layout.fillMaxHeight
import androidx.compose.foundation.layout.height
import androidx.compose.foundation.shape.CircleShape
import androidx.compose.material3.MaterialTheme
import androidx.compose.runtime.Composable
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.clip
import androidx.compose.ui.graphics.Brush
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.unit.dp

/** Cold-to-warm gradient used by the daily temperature range bars. */
private val temperatureGradient = listOf(Color(0xFF4FC3F7), Color(0xFFFFB74D))

/** A single-degree day would otherwise collapse to an invisible bar. */
private const val MIN_FILL_FRACTION = 0.05f

/**
 * Horizontal bar showing where [low]..[high] sits inside the [scaleMin]..[scaleMax] range
 * shared by every day of the forecast.
 */
@Composable
internal fun TemperatureRangeBar(
    low: Double,
    high: Double,
    scaleMin: Double,
    scaleMax: Double,
    modifier: Modifier = Modifier,
) {
    val scaleSpan = (scaleMax - scaleMin).takeIf { it > 0.0 } ?: 1.0
    val startFraction = ((low - scaleMin) / scaleSpan).toFloat().coerceIn(0f, 1f)
    val endFraction = ((high - scaleMin) / scaleSpan).toFloat().coerceIn(startFraction, 1f)
    val fillFraction = (endFraction - startFraction).coerceAtLeast(MIN_FILL_FRACTION)
    val leadingFraction = startFraction.coerceAtMost(1f - fillFraction)
    val trailingFraction = 1f - leadingFraction - fillFraction

    Row(
        modifier =
            modifier
                .height(8.dp)
                .clip(CircleShape)
                .background(MaterialTheme.colorScheme.surfaceVariant),
    ) {
        if (leadingFraction > 0f) {
            Spacer(modifier = Modifier.weight(leadingFraction))
        }
        Box(
            modifier =
                Modifier
                    .weight(fillFraction)
                    .fillMaxHeight()
                    .clip(CircleShape)
                    .background(Brush.horizontalGradient(temperatureGradient)),
        )
        if (trailingFraction > 0f) {
            Spacer(modifier = Modifier.weight(trailingFraction))
        }
    }
}
